File: samples/java/DeliveryDispatch/Server/Tracking/src/main/java/systems/zlink/samples/deliverydispatch/server/tracking/TrackingServerApplication.java
package systems.zlink.samples.deliverydispatch.server.tracking;

import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter;
import systems.zlink.samples.deliverydispatch.server.configuration.EvidenceStore;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleApplication;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleLocationStore;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleTopology;

import java.net.URI;

@EnableZLinkFramework
@EnableConfigurationProperties(SampleTopology.class)
@SpringBootApplication(
        proxyBeanMethods = false,
        scanBasePackageClasses = TrackingServerApplication.class)
public final class TrackingServerApplication {
    private TrackingServerApplication() {}

    public static AutoCloseable run(String configPath) {
        return SampleApplication.start(TrackingServerApplication.class, configPath)::close;
    }

    @Bean
    ZLinkFrameworkConfigurer trackingFramework(SampleTopology topology) {
        return options -> {
            options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL);
            options.addHandlersFromPackageOf(TrackingServerApplication.class);
            ZLinkMeshNodeBuilder trackingSpot =
                    options.addRouteMesh(SampleNames.CustomerSpotDiscovery);
            trackingSpot
                    .listen(topology.trackingSpotEndpoint())
                    .setRoutingId(RoutingId.from(SampleNames.TrackingNode));
            trackingSpot.objects().client();
            URI trackingEndpoint = URI.create(topology.trackingChannelEndpoint());
            options.addClientServerChannel(SampleNames.TrackingChannel)
                    .server()
                    .setBindHost(trackingEndpoint.getHost())
                    .setAdvertiseHost(trackingEndpoint.getHost())
                    .listen(trackingEndpoint.getPort())
                    .addHandlerGroup("tracking");
        };
    }

    @Bean(destroyMethod = "close")
    ZLinkRedisLocationStore locationStore(SampleTopology topology) {
        return SampleLocationStore.create(topology);
    }

    @Bean
    DeliveryDispatchReadinessReporter readinessReporter(ZLinkRouteMeshRuntime routeMeshRuntime) {
        return new DeliveryDispatchReadinessReporter(
                routeMeshRuntime, SampleNames.TrackingNode, SampleNames.CustomerSpotDiscovery);
    }

    @Bean
    EvidenceStore evidenceStore() {
        return new EvidenceStore();
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Tracking/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/tracking/TrackingServerApplication.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.tracking

import java.net.URI
import kotlinx.coroutines.Dispatchers
import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.annotation.Bean
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryEvidenceStore
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleLocationStore
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleTopology

@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = [TrackingServerApplication::class],
)
class TrackingServerApplication {
    @Bean
    fun trackingFramework(): ZLinkFrameworkConfigurer = ZLinkFrameworkConfigurer { options ->
        options.useCoroutineHandlers(Dispatchers.Default)
        options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL)

        options.addHandlersFromPackageOf(TrackingServerApplication::class.java)
        val trackingSpot =
            options
                .addRouteMesh(SampleNames.CustomerSpotMesh)
                .listen(SampleTopology.TrackingSpotEndpoint)
                .setRoutingIdPrefix("delivery-tracking")
        trackingSpot.objects().client()
        val trackingEndpoint = URI.create(SampleTopology.TrackingChannelEndpoint)
        options
            .addClientServerChannel(SampleNames.TrackingChannel)
            .server()
            .setBindHost(trackingEndpoint.host)
            .setAdvertiseHost(trackingEndpoint.host)
            .listen(trackingEndpoint.port)
            .addHandlerGroup("tracking")
    }

    @Bean fun locationStore(): ZLinkRedisLocationStore = SampleLocationStore.create()

    @Bean(destroyMethod = "close")
    fun readinessReporter(meshes: ZLinkRouteMeshRuntime): DeliveryDispatchReadinessReporter =
        DeliveryDispatchReadinessReporter.route(
            SampleNames.TrackingNode,
            SampleNames.CustomerSpotMesh,
            meshes,
        )

    @Bean
    fun evidenceStore(): DeliveryEvidenceStore =
        DeliveryEvidenceStore(SampleTopology.StateDirectory)

    companion object {
        fun run(args: Array<String> = emptyArray()): AutoCloseable {
            val builder =
                SpringApplicationBuilder(TrackingServerApplication::class.java)
                    .web(WebApplicationType.NONE)
            builder.application().setKeepAlive(true)
            val context = builder.run(*args)
            return AutoCloseable { context.close() }
        }
    }
}
